### DISCUSSION TITLE
Squashing migrations into one creates an impossibility to create any follow up migrations

### DISCUSSION THREAD

**Discussion by georges-ryan** [prisma/prisma#25370] [Q&A]
### Bug description

Good morning 
We ha ve a project with ~650 migration files in our prisma project. 
We want to squash them into one to clean our history with only one file. 
Once all these steps [here](https://www.prisma.io/docs/orm/prisma-migrate/workflows/squashing-migrations#how-to-create-a-clean-history-in-a-production-environment) are done, we end up with the wanted squashed migration into a single file. Yet, any follow up migration will fail with the following error 
```
The following migration(s) are applied to the database but missing from the local migrations directory (...)
```
that can be fixed by deleting all history in prisma table (keeping only the squashed migration of course). Not sure this is the right way.

Note: I'm testing all locally for now.  

### How to reproduce

Following prisma documentation [here](https://www.prisma.io/docs/orm/prisma-migrate/workflows/squashing-migrations#how-to-create-a-clean-history-in-a-production-environment)

- clean all `migrations` content` 
- run the following command to create a new migration file
```
npx dotenv -c development -- prisma migrate diff \
 --from-empty \
 --to-schema-datamodel ./prisma/schema.prisma \
 --script > ./prisma/migrations/000000000000_squashed_migrations/migration.sql
```
- mark migration as resolved 
```
npx prisma migrate resolve  --applied 000000000000_squashed_migrations
```

We end up with one migrations folder `000000000000_squashed_migrations` with a good representation of the schema and the history still exists on `_prisma_migrations` table. 

Now let's say I want to add a migration to my schema and run it using 
```
npx dotenv -c development -- prisma migrate dev --name some-migration
```
I end up with the following error (Note that I shorten it because, as I said, ~650 migration folders were deleted)
```
The following migration(s) are applied to the database but missing from the local migrations directory: 20220822160247_init, 20220825104947_add_bridge_event_unique_fields, 20220825110128_add_bridge_event_status, 20220830144555_add_user_verification_status, 20220901072729_rename_verification_status_enum, 20220831160011_add_card_id_to_transactions_table, 20220914153021_add_account_reference, 20220915152026_account_reference_required, 20220920093423_transaction_reason, 20220926144846_missing_defacto_event_type, 20220928130205_add_external_reference_created_and_updated_at, 20220928130253_remove_default_on_external_reference_updated_at, 20220929093955_supplier_payout_loan_category, 20221007134346_delete_status_defacto_events, 20221010143557(...)
```

Now I fixed it by deleting all previous migrations from the prisma table but it seems a bit brutal and it was not specified to do it in the doc. 

One more question about the `resolve` command 
```
npx dotenv -c development -- prisma migrate resolve --applied 000000000000_squashed_migrations
 ```
 does not pass the `applied_step_counts` to 1, is this normal ? 
![image](https://github.com/user-attachments/assets/85947ad4-5d69-42eb-876b-e84da020eec7)

Thank you for your help, if you need any further details, I'd, be glad to share.

### Expected behavior

I want my migrations to be squashed into one, keeping the history in my prisma table and be able to run further migration. 

### Prisma information

I'm not sure I'm allowed to share the model... 

### Environment & setup

- Database: PostgreSQL
- Node.js version: v20.12.1

### Prisma Version

```
5.19.1
```

**Comment by felixizraitel** [comment]
I encountered the same issue. Ended up backing up the migrations table and deleting the records to get this to work.

**Reply by georges-ryan** [reply]
Thank you, yeah it was a solution we thought off but it seems weird with to many manual moves.

**Reply by georges-ryan** [reply]
Plus it's not specified, I wonder if they changed anything after writing the docs

**Reply by CaitW** [reply]
This is also what I did. I deleted every entry except the last (the squashed migration) and everything ran as expected.

**Comment by ludralph** [comment]
Hi @georges-ryan 👋 

The error indicates that there are migrations applied to your database that are not present in your local migrations directory. This can happen after squashing migrations if the migration history is not properly aligned between your local setup and the production database.

You can use the `migrate resolve` command to mark the squashed migration as applied in your production database. This will help align your local migration history with the production database state. This process is also known as [Baselining](https://www.prisma.io/docs/orm/prisma-migrate/workflows/baselining#baselining-a-database). 
```
npx prisma migrate resolve --applied <name-of-the-squashed-migration>
```
Replace <name-of-the-squashed-migration> with the actual name of your squashed migration directory. This step ensures that Prisma knows the migration has already been applied to the database, even though it's not present in your local migrations directory.

If you encounter errors about missing migrations, you may need to manually delete old migration entries from the `_prisma_migrations` table, leaving only the entry for the squashed migration.

The `applied_steps_count` in the `_prisma_migrations` table is intended to reflect the number of migration steps that have been successfully applied for a given migration. It typically starts at 0 and should increment to 1 (or higher, depending on the number of steps in the migration) once the migration is successfully applied. If the `applied_steps_count` remains 0, it might be due to a problem with the migration process.

**Reply by georges-ryan** [reply]
Hello 
Thank you for your reply 

> You can use the migrate resolve command to mark the squashed migration as applied in your production database. This will help align your local migration history with the production database state. This process is also known as [Baselining](https://www.prisma.io/docs/orm/prisma-migrate/workflows/baselining#baselining-a-database).

I did this. And yes I had to remove all history for the `_prisma_migrations` table. It was working this way. 
But is deleting all data coming from views and data inserted in the migration files. Reproducing a prod-like database on development or local (non-prod) environment becomes impossible.

If I want to start fresh from an empty locally, I would lose all data if I run my migration after squashing them. It appears that those are not taken into consideration.
Would you have any recommandation for this ?

**Reply by ludralph** [reply]
Hi @georges-ryan 
Some recommendations would be:

1. Custom SQL in Migrations: If you need to preserve specific SQL operations (like creating views) that were in your original migrations, you can add these to your squashed migration file manually. The [Prisma documentation on customizing migrations](https://www.prisma.io/docs/orm/prisma-migrate/workflows/customizing-migrations) explains how to do this.

2. Baselining with Data Preservation: When baselining your database (which is what you did with migrate resolve), you can take an additional step to preserve your data. Before removing the migration history, you can:
a. Take a full backup of your database.
b. After baselining and applying the squashed migration, restore the data from your backup.
This approach is mentioned in a [GitHub comment](https://github.com/prisma/prisma/issues/11668#issuecomment-1033743782) as a workaround for preserving data while resolving migration issues.

3. Separate Schema Changes and Data Migrations: It's generally a good practice to separate schema changes from data migrations. This allows you to squash schema migrations without losing data insertions. You can keep your data migrations in separate files that run after the schema migrations.

**Comment by ludralph** [comment]
Hi there,

To keep our discussions organized and focused on the most relevant topics, we’re reviewing and tidying up our backlog. As part of this process, we’re closing discussions that haven’t had any recent activity and appear to be outdated.

If this discussion is still important to you or unresolved, we’d love to hear from you! Feel free to reopen it or start a new one with updated details.

For more details about our priorities and vision for the future of Prisma ORM, check out our latest blog post: https://www.prisma.io/blog/prisma-orm-manifesto.

Thank you for your understanding and being part of the community!;
